File: Example/javaapp/src/main/java/com/stripe/example/javaapp/viewmodel/UpdateReaderViewModel.java
package com.stripe.example.javaapp.viewmodel;

import android.app.Application;
import android.content.Context;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import com.stripe.example.javaapp.R;
import com.stripe.stripeterminal.callable.Cancelable;
import com.stripe.stripeterminal.model.external.Reader;
import com.stripe.stripeterminal.model.external.ReaderSoftwareUpdate;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class UpdateReaderViewModel extends AndroidViewModel {
    @NotNull public MutableLiveData<Double> progress;
    @NotNull public MutableLiveData<Boolean> hasStartedFetchingUpdate;
    @NotNull public MutableLiveData<Boolean> hasFinishedFetchingUpdate;
    @NotNull public MutableLiveData<Boolean> hasStartedInstallingUpdate;
    @NotNull public MutableLiveData<Boolean> hasFinishedInstallingUpdate;
    @NotNull public MutableLiveData<@Nullable ReaderSoftwareUpdate> readerSoftwareUpdate;

    @NotNull public MediatorLiveData<Boolean> checkForUpdateDescriptionVisibility = new MediatorLiveData<>();
    @NotNull public MediatorLiveData<Boolean> checkForUpdateButtonVisibility = new MediatorLiveData<>();
    @NotNull public MediatorLiveData<Integer> checkForUpdateButtonColor = new MediatorLiveData<>();
    @NotNull public MediatorLiveData<Integer> checkForUpdateButtonText = new MediatorLiveData<>();
    @NotNull public MediatorLiveData<String> checkForUpdateDescriptionText = new MediatorLiveData<>();
    @NotNull public MediatorLiveData<Boolean> doneButtonVisibility = new MediatorLiveData<>();
    @NotNull public MediatorLiveData<Boolean> installDisclaimerVisibility = new MediatorLiveData<>();

    @Nullable public Cancelable fetchUpdateOperation = null;
    @Nullable public Cancelable installOperation = null;
    @Nullable public Reader reader = null;

    public UpdateReaderViewModel(@NotNull Application app) {
        super(app);

        progress = new MutableLiveData<>(0.0);
        hasStartedFetchingUpdate = new MutableLiveData<>(false);
        hasFinishedFetchingUpdate = new MutableLiveData<>(false);
        hasStartedInstallingUpdate = new MutableLiveData<>(false);
        hasFinishedInstallingUpdate = new MutableLiveData<>(false);
        readerSoftwareUpdate = new MutableLiveData<>(null);

        final Observer<Object> updateDescriptionVisibilityLambda = value ->
                checkForUpdateDescriptionVisibility.setValue(getCheckForUpdateDescriptionVisibility());
        checkForUpdateDescriptionVisibility.addSource(hasStartedFetchingUpdate, updateDescriptionVisibilityLambda);
        checkForUpdateDescriptionVisibility.addSource(hasFinishedFetchingUpdate, updateDescriptionVisibilityLambda);
        checkForUpdateDescriptionVisibility.addSource(hasStartedInstallingUpdate, updateDescriptionVisibilityLambda);
        checkForUpdateDescriptionVisibility.addSource(readerSoftwareUpdate, updateDescriptionVisibilityLambda);

        final Observer<Object> checkForUpdateButtonVisibilityLambda = value ->
                checkForUpdateButtonVisibility.setValue(getCheckForUpdateButtonVisibility());
        checkForUpdateButtonVisibility.addSource(hasStartedInstallingUpdate, checkForUpdateButtonVisibilityLambda);
        checkForUpdateButtonVisibility.addSource(hasFinishedInstallingUpdate, checkForUpdateButtonVisibilityLambda);
        checkForUpdateButtonVisibility.addSource(hasStartedFetchingUpdate, checkForUpdateButtonVisibilityLambda);
        checkForUpdateButtonVisibility.addSource(hasFinishedFetchingUpdate, checkForUpdateButtonVisibilityLambda);
        checkForUpdateButtonVisibility.addSource(readerSoftwareUpdate, checkForUpdateButtonVisibilityLambda);

        final Observer<Boolean> checkForUpdateButtonColorLambda =
                value -> checkForUpdateButtonColor.setValue(getCheckForUpdateButtonColor());
        checkForUpdateButtonColor.addSource(hasStartedFetchingUpdate, checkForUpdateButtonColorLambda);
        checkForUpdateButtonColor.addSource(hasFinishedFetchingUpdate, checkForUpdateButtonColorLambda);

        final Observer<Object> checkForUpdateButtonTextLambda =
                value -> checkForUpdateButtonText.setValue(getCheckForUpdateButtonText());
        checkForUpdateButtonText.addSource(hasStartedFetchingUpdate, checkForUpdateButtonTextLambda);
        checkForUpdateButtonText.addSource(hasFinishedFetchingUpdate, checkForUpdateButtonTextLambda);
        checkForUpdateButtonText.addSource(hasStartedInstallingUpdate, checkForUpdateButtonTextLambda);
        checkForUpdateButtonText.addSource(readerSoftwareUpdate, checkForUpdateButtonTextLambda);

        final Observer<Object> checkForUpdateDescriptionTextLambda =
                value -> checkForUpdateDescriptionText.setValue(getCheckForUpdateDescriptionText());
        checkForUpdateDescriptionText.addSource(hasStartedFetchingUpdate, checkForUpdateDescriptionTextLambda);
        checkForUpdateDescriptionText.addSource(hasFinishedFetchingUpdate, checkForUpdateDescriptionTextLambda);
        checkForUpdateDescriptionText.addSource(hasStartedInstallingUpdate, checkForUpdateDescriptionTextLambda);
        checkForUpdateDescriptionText.addSource(hasFinishedInstallingUpdate, checkForUpdateDescriptionTextLambda);
        checkForUpdateDescriptionText.addSource(progress, checkForUpdateDescriptionTextLambda);

        final Observer<Object> doneButtonVisibilityLambda =
                value -> doneButtonVisibility.setValue(getDoneButtonVisibility());
        doneButtonVisibility.addSource(hasFinishedInstallingUpdate, doneButtonVisibilityLambda);
        doneButtonVisibility.addSource(hasStartedFetchingUpdate, doneButtonVisibilityLambda);
        doneButtonVisibility.addSource(hasFinishedFetchingUpdate, doneButtonVisibilityLambda);
        doneButtonVisibility.addSource(hasStartedInstallingUpdate, doneButtonVisibilityLambda);
        doneButtonVisibility.addSource(readerSoftwareUpdate, doneButtonVisibilityLambda);

        final Observer<Object> installDisclaimerVisibilityLambda =
                value -> installDisclaimerVisibility.setValue(getInstallDisclaimerVisiblity());
        installDisclaimerVisibility.addSource(hasStartedInstallingUpdate, installDisclaimerVisibilityLambda);
        installDisclaimerVisibility.addSource(hasFinishedFetchingUpdate, installDisclaimerVisibilityLambda);
    }

    private boolean getCheckForUpdateDescriptionVisibility() {
        return (hasStartedFetchingUpdate.getValue() && !hasFinishedFetchingUpdate.getValue()) ||
                (hasStartedFetchingUpdate.getValue() && hasFinishedFetchingUpdate.getValue() &&
                !hasStartedInstallingUpdate.getValue() && readerSoftwareUpdate.getValue() != null) ||
                hasStartedInstallingUpdate.getValue();
    }

    private boolean getCheckForUpdateButtonVisibility() {
        return !(hasStartedInstallingUpdate.getValue() && hasFinishedInstallingUpdate.getValue()) ||
                (hasStartedFetchingUpdate.getValue() && hasFinishedFetchingUpdate.getValue() &&
                !hasStartedInstallingUpdate.getValue() && readerSoftwareUpdate.getValue() != null);
    }

    private int getCheckForUpdateButtonColor() {
        if (hasStartedFetchingUpdate.getValue() && !hasFinishedFetchingUpdate.getValue()) {
            return R.color.colorPrimaryDark;
        } else {
            return R.color.colorAccent;
        }
    }

    private int getCheckForUpdateButtonText() {
        if (hasStartedFetchingUpdate.getValue() && !hasFinishedFetchingUpdate.getValue()) {
            return R.string.checking_for_update;
        } else if (hasStartedFetchingUpdate.getValue() && hasFinishedFetchingUpdate.getValue() &&
                !hasStartedInstallingUpdate.getValue()) {
            return readerSoftwareUpdate.getValue() != null ?
                    R.string.install_update : R.string.no_update_available;
        } else if (hasStartedInstallingUpdate.getValue()) {
            return R.string.update_in_progress;
        } else {
            return R.string.check_for_update;
        }
    }

    @NotNull
    private String getCheckForUpdateDescriptionText() {
        final Context context = getApplication().getApplicationContext();
        if (hasStartedFetchingUpdate.getValue() && !hasFinishedFetchingUpdate.getValue()) {
            return context.getString(R.string.checking_for_update);
        } else if (hasStartedFetchingUpdate.getValue() && hasFinishedFetchingUpdate.getValue() &&
                !hasStartedInstallingUpdate.getValue() && readerSoftwareUpdate.getValue() != null) {
            return context.getString(R.string.install_explanation,
                    readerSoftwareUpdate.getValue().getVersion(),
                    readerSoftwareUpdate.getValue().getTimeEstimate().getDescription());
        } else if (hasStartedInstallingUpdate.getValue()) {
            return hasFinishedInstallingUpdate.getValue() ? context.getString(R.string.update_complete) :
                context.getString(R.string.update_progress, Double.toString(
                        Math.round((progress.getValue() != null ? progress.getValue() : 0.0) * 100)));
        } else {
            return context.getString(R.string.update_explanation);
        }
    }

    private boolean getDoneButtonVisibility() {
        return hasFinishedInstallingUpdate.getValue() || (hasStartedFetchingUpdate.getValue() &&
                hasFinishedFetchingUpdate.getValue() && !hasStartedInstallingUpdate.getValue() &&
                readerSoftwareUpdate.getValue() == null);
    }

    private boolean getInstallDisclaimerVisiblity() {
        return hasStartedInstallingUpdate.getValue() && hasFinishedFetchingUpdate.getValue();
    }
}

File: Example/kotlinapp/src/main/java/com/stripe/example/viewmodel/UpdateReaderViewModel.kt
package com.stripe.example.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.stripe.example.R
import com.stripe.example.StripeTerminalApplication
import com.stripe.stripeterminal.callable.Cancelable
import com.stripe.stripeterminal.model.external.Reader
import com.stripe.stripeterminal.model.external.ReaderSoftwareUpdate
import kotlin.math.roundToInt

class UpdateReaderViewModel
constructor(application: Application) : AndroidViewModel(application) {
    var progress: MutableLiveData<Double> = MutableLiveData(0.0)
    var hasStartedFetchingUpdate: MutableLiveData<Boolean> = MutableLiveData(false)
    var hasFinishedFetchingUpdate: MutableLiveData<Boolean> = MutableLiveData(false)
    var hasStartedInstallingUpdate: MutableLiveData<Boolean> = MutableLiveData(false)
    var hasFinishedInstallingUpdate: MutableLiveData<Boolean> = MutableLiveData(false)
    var readerSoftwareUpdate: MutableLiveData<ReaderSoftwareUpdate?> = MutableLiveData(null)
    var fetchUpdateOperation: Cancelable? = null
    var installOperation: Cancelable? = null
    var reader: Reader? = null

    val checkForUpdateDescriptionVisibility: MediatorLiveData<Boolean> = MediatorLiveData()
    val checkForUpdateButtonVisibility: MediatorLiveData<Boolean> = MediatorLiveData()
    val checkForUpdateButtonColor: MediatorLiveData<Int> = MediatorLiveData()
    val checkForUpdateButtonText: MediatorLiveData<Int> = MediatorLiveData()
    val checkForUpdateDescriptionText: MediatorLiveData<String> = MediatorLiveData()
    val doneButtonVisibility: MediatorLiveData<Boolean> = MediatorLiveData()

    init {
        val updateDescriptionVisibilityLambda: Observer<Boolean> = Observer {
            checkForUpdateDescriptionVisibility.value = getCheckForUpdateDescriptionVisibility()
        }
        checkForUpdateDescriptionVisibility
                .addSource(hasStartedFetchingUpdate, updateDescriptionVisibilityLambda)
        checkForUpdateDescriptionVisibility
                .addSource(hasFinishedFetchingUpdate, updateDescriptionVisibilityLambda)
        checkForUpdateDescriptionVisibility
                .addSource(hasStartedInstallingUpdate, updateDescriptionVisibilityLambda)
        checkForUpdateDescriptionVisibility.addSource(readerSoftwareUpdate) {
            checkForUpdateDescriptionVisibility.value = getCheckForUpdateDescriptionVisibility()
        }

        val checkForUpdateButtonVisibilityLambda: Observer<Boolean> = Observer {
            checkForUpdateButtonVisibility.value = getCheckForUpdateButtonVisibility()
        }
        checkForUpdateButtonVisibility
                .addSource(hasStartedInstallingUpdate, checkForUpdateButtonVisibilityLambda)
        checkForUpdateButtonVisibility
                .addSource(hasFinishedInstallingUpdate, checkForUpdateButtonVisibilityLambda)
        checkForUpdateButtonVisibility
                .addSource(hasStartedFetchingUpdate, checkForUpdateButtonVisibilityLambda)
        checkForUpdateButtonVisibility
                .addSource(hasFinishedFetchingUpdate, checkForUpdateButtonVisibilityLambda)
        checkForUpdateButtonVisibility.addSource(readerSoftwareUpdate) {
            checkForUpdateButtonVisibility.value = getCheckForUpdateButtonVisibility()
        }

        val checkForUpdateButtonColorLambda = Observer<Boolean> {
            checkForUpdateButtonColor.value = getCheckForUpdateButtonColor()
        }
        checkForUpdateButtonColor
                .addSource(hasStartedFetchingUpdate, checkForUpdateButtonColorLambda)
        checkForUpdateButtonColor
                .addSource(hasFinishedFetchingUpdate, checkForUpdateButtonColorLambda)

        val checkForUpdateButtonTextLambda = Observer<Boolean> {
            checkForUpdateButtonText.value = getCheckForUpdateButtonText()
        }
        checkForUpdateButtonText
                .addSource(hasStartedFetchingUpdate, checkForUpdateButtonTextLambda)
        checkForUpdateButtonText
                .addSource(hasFinishedFetchingUpdate, checkForUpdateButtonTextLambda)
        checkForUpdateButtonText
                .addSource(hasStartedInstallingUpdate, checkForUpdateButtonTextLambda)
        checkForUpdateButtonText.addSource(readerSoftwareUpdate) {
            checkForUpdateButtonText.value = getCheckForUpdateButtonText()
        }

        val checkForUpdateDescriptionTextLambda = Observer<Any> {
            checkForUpdateDescriptionText.value = getCheckForUpdateDescriptionText()
        }
        checkForUpdateDescriptionText.addSource(hasStartedFetchingUpdate, checkForUpdateDescriptionTextLambda)
        checkForUpdateDescriptionText.addSource(hasFinishedFetchingUpdate, checkForUpdateDescriptionTextLambda)
        checkForUpdateDescriptionText.addSource(hasStartedInstallingUpdate, checkForUpdateDescriptionTextLambda)
        checkForUpdateDescriptionText.addSource(hasFinishedInstallingUpdate, checkForUpdateDescriptionTextLambda)
        checkForUpdateDescriptionText.addSource(progress, checkForUpdateDescriptionTextLambda)

        val doneButtonVisibilityLambda = Observer<Boolean> {
            doneButtonVisibility.value = getDoneButtonVisibility()
        }
        doneButtonVisibility.addSource(hasFinishedInstallingUpdate, doneButtonVisibilityLambda)
        doneButtonVisibility.addSource(hasStartedFetchingUpdate, doneButtonVisibilityLambda)
        doneButtonVisibility.addSource(hasFinishedFetchingUpdate, doneButtonVisibilityLambda)
        doneButtonVisibility.addSource(hasStartedInstallingUpdate, doneButtonVisibilityLambda)
        doneButtonVisibility.addSource(readerSoftwareUpdate) {
            doneButtonVisibility.value = getDoneButtonVisibility()
        }
    }

    private fun getCheckForUpdateDescriptionVisibility(): Boolean {
        return (hasStartedFetchingUpdate.value!! && !hasFinishedFetchingUpdate.value!!) ||
                (hasStartedFetchingUpdate.value!! && hasFinishedFetchingUpdate.value!! &&
                !hasStartedInstallingUpdate.value!! && readerSoftwareUpdate.value != null) ||
                hasStartedInstallingUpdate.value!!
    }

    private fun getCheckForUpdateButtonVisibility(): Boolean {
        return !(hasStartedInstallingUpdate.value!! && hasFinishedInstallingUpdate.value!!) ||
                (hasStartedFetchingUpdate.value!! && hasFinishedFetchingUpdate.value!! &&
                        !hasStartedInstallingUpdate.value!! && readerSoftwareUpdate.value != null)
    }

    private fun getCheckForUpdateButtonColor(): Int {
        return if (hasStartedFetchingUpdate.value!! && !hasFinishedFetchingUpdate.value!!) {
            R.color.colorPrimaryDark
        } else {
            R.color.colorAccent
        }
    }

    private fun getCheckForUpdateButtonText(): Int {
        return if (hasStartedFetchingUpdate.value!! && !hasFinishedFetchingUpdate.value!!) {
            R.string.checking_for_update
        } else if (hasStartedFetchingUpdate.value!! && hasFinishedFetchingUpdate.value!! &&
                !hasStartedInstallingUpdate.value!!) {
            if (readerSoftwareUpdate.value != null) R.string.install_update else R.string.no_update_available
        } else if (hasStartedInstallingUpdate.value!!) {
            R.string.update_in_progress
        } else {
            R.string.check_for_update
        }
    }

    private fun getCheckForUpdateDescriptionText(): String {
        val context = getApplication<StripeTerminalApplication>().applicationContext
        return if (hasStartedFetchingUpdate.value!! && !hasFinishedFetchingUpdate.value!!) {
            context.getString(R.string.checking_for_update)
        } else if (hasStartedFetchingUpdate.value!! && hasFinishedFetchingUpdate.value!! &&
                !hasStartedInstallingUpdate.value!! && readerSoftwareUpdate.value != null) {
            context.getString(R.string.install_explanation,
                    readerSoftwareUpdate.value!!.version,
                    readerSoftwareUpdate.value!!.timeEstimate.description)
        } else if (hasStartedInstallingUpdate.value!!) {
            if (hasFinishedInstallingUpdate.value!!) context.getString(R.string.update_complete)
            else context.getString(R.string.update_progress, (((progress.value ?: 0.0) * 100).roundToInt()).toString())
        } else {
            context.getString(R.string.update_explanation)
        }
    }

    private fun getDoneButtonVisibility(): Boolean {
        return hasFinishedInstallingUpdate.value!! || (hasStartedFetchingUpdate.value!! &&
                hasFinishedFetchingUpdate.value!! && !hasStartedInstallingUpdate.value!! &&
                readerSoftwareUpdate.value == null)
    }
}
